File: src/games/stendhal/client/entity/Door.java
/* $Id$ */
/***************************************************************************
 *                      (C) Copyright 2003 - Marauroa                      *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.client.entity;

import marauroa.common.game.RPObject;

/**
 * A door entity.
 */
public class Door extends Portal {
	/**
	 * Open state property.
	 */
	public static final Property PROP_OPEN = new Property();

	/**
	 * Whether the door is open.
	 */
	private boolean open;

	/**
	 * Create a door entity.
	 */
	public Door() {
		// default constructor
	}

	//
	// Door
	//

	/**
	 * Check if the door is open.
	 *
	 * @return <code>true</code> if the door is open.
	 */
	public boolean isOpen() {
		return open;
	}

	//
	// Entity
	//

	/**
	 * Initialize this entity for an object.
	 *
	 * @param object
	 *            The object.
	 *
	 * @see #release()
	 */
	@Override
	public void initialize(final RPObject object) {
		super.initialize(object);

		/*
		 * Open state
		 */
		open = object.has("open");
	}

	//
	// RPObjectChangeListener
	//

	/**
	 * The object added/changed attribute(s).
	 *
	 * @param object
	 *            The base object.
	 * @param changes
	 *            The changes.
	 */
	@Override
	public void onChangedAdded(final RPObject object, final RPObject changes) {
		super.onChangedAdded(object, changes);

		/*
		 * Open state
		 */
		if (changes.has("open")) {
			open = true;
			fireChange(PROP_OPEN);
		}
	}

	/**
	 * The object removed attribute(s).
	 *
	 * @param object
	 *            The base object.
	 * @param changes
	 *            The changes.
	 */
	@Override
	public void onChangedRemoved(final RPObject object, final RPObject changes) {
		super.onChangedRemoved(object, changes);

		/*
		 * Open state
		 */
		if (changes.has("open")) {
			open = false;
			fireChange(PROP_OPEN);
		}
	}
}

File: src/js/stendhal/entity/Door.ts
/***************************************************************************
 *                   (C) Copyright 2003-2024 - Stendhal                    *
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

import { RenderingContext2D } from "util/Types";
import { MenuItem } from "../action/MenuItem";
import { Portal } from "./Portal";
import { Paths } from "../data/Paths";
import { ImageSprite } from "sprite/image/ImageSprite";
import { images } from "sprite/image/ImageManager";

export class Door extends Portal {

	override zIndex = 5000;

	override set(key: string, value: any) {
		super.set(key, value);
		if (key === "class") {
			if (this.imageSprite) {
				this.imageSprite?.free();
				this.imageSprite = new ImageSprite(
					images.load(Paths.sprites + "/doors/" + value + ".png"),
					0, 0, 0, 0);
			}
		}
	}

	override draw(ctx: RenderingContext2D) {
		let image = this.imageSprite?.imageRef?.image;
		if (!image) {
			return;
		}
		let height = image.height / 2;
		let x = (this["x"] * 32) - ((image.width - 32) / 2);
		let y = (this["y"] * 32) - ((height - 32) / 2);

		let offsetY = height;
		if (this["open"] === "") {
			offsetY = 0;
		}
		ctx.drawImage(image, 0, offsetY, image.width, height, x, y, image.width, height);
	}

	override buildActions(list: MenuItem[]) {
		list.push({
			title: "Look",
			type: "look"
		});
		list.push({
			title: "Use",
			type: "use"
		});
	}

	override isVisibleToAction(_filter: boolean) {
		return true;
	}

	override getCursor(_x: number, _y: number) {
		return "url(" + Paths.sprites + "/cursor/portal.png) 1 3, auto";
	}

}
